### ISSUE TITLE
Config to reproduce Parallelized DeltaNet results

### ISSUE THREAD

**Issue by kazuki-irie** [fla-org/flame#30]
Hello flame authors,

First of all, thanks so much for this great project! This looks very promising!

I've seen a couple of example configs in the README, but I think it would be useful to include training configs that (roughly) reproduce results from prior papers.

For example, I'm interested in (roughly) reproducing the results of 340M-param 15B-token DeltaNet from https://arxiv.org/abs/2406.06484 (ideally using the same subsets of SlimPajama used in the paper).

What should be the full config for this? 
I did see the model config in https://github.com/fla-org/flame/blob/main/configs/delta_net_340M.json, but could you provide hints on how the training config should be? (like here, https://github.com/fla-org/flame?tab=readme-ov-file#training-recipes)

Also, currently, when I use flame as-is to train DeltaNet, I run into `ChunkDeltaRuleFunction does not support float32. Please use bfloat16.` So, to enable training of DeltaNet (and other models with the same requirement), users need to modify flame to add autocasting.

**Comment by kazuki-irie** [comment]
Closing as I was able to figure it out myself.
I plan to share the configs/settings once I complete the full set of experiments, in case it's helpful for others in the future.

**Comment by yzhangcs** [comment]
> Also, currently, when I use flame as-is to train DeltaNet, I run into ChunkDeltaRuleFunction does not support float32. Please use bfloat16. So, to enable training of DeltaNet (and other models with the same requirement), users need to modify flame to add autocasting.

Hi, I can't reproduce this problem, did you set `NGPU` properly?

**Comment by yzhangcs** [comment]
> I did see the model config in https://github.com/fla-org/flame/blob/main/configs/delta_net_340M.json, but could you provide hints on how the training config should be? (like here, https://github.com/fla-org/flame?tab=readme-ov-file#training-recipes)

The configurations listed in [https://github.com/fla-org/flame?tab=readme-ov-file#training-recipes](https://www.google.com/url?sa=E&q=https%3A%2F%2Fgithub.com%2Ffla-org%2Fflame%3Ftab%3Dreadme-ov-file%23training-recipes) were suggested by @sustcsonglin. Following her suggestions to improve the somewhat suboptimal original settings, we adopted a larger learning rate. We also uniformly set the default training mode to varlen, as it is supported by fla for most models.

I do suggest you conduct some exps to find the optimal settings yourself. Very glad to hear your feedbacks.

**Comment by kazuki-irie** [comment]
Hi Yu,

Thanks for your response.

> Hi, I can't reproduce this problem, did you set NGPU properly?

I believe so, yes. For now, I fixed this by explicitly adding `torch.autocast` inside flame (which, I guess, is not how I'm supposed to fix it). But let's hold off on this for now. I will report back soon once I've completed all my current experimentations.

Thanks again for this very useful toolkit.

**Comment by yzhangcs** [comment]
@kazuki-irie Could you paste the running cmd here for your deltanet models

**Comment by kazuki-irie** [comment]
```
export NGPU=4
export CUDA_VISIBLE_DEVICES=0,1,2,3

MODEL_CONFIG=configs/delta_net_340M.json
TOKENIZER=fla-hub/transformer-1.3B-100B
DUMP=exp/fineweb-delta-32

DATASET=HuggingFaceFW/fineweb-edu
DATASET_NAME=sample-100BT

STEPS=28672

bash train.sh \
  --job.config_file flame/models/fla.toml \
  --job.dump_folder ${DUMP} \
  --model.config ${MODEL_CONFIG} \
  --model.tokenizer_path ${TOKENIZER} \
  --optimizer.name AdamW \
  --optimizer.eps 1e-15 \
  --optimizer.lr 1e-3 \
  --lr_scheduler.warmup_steps 1024 \
  --lr_scheduler.lr_min 0.1 \
  --lr_scheduler.decay_type cosine \
  --training.batch_size 32 \
  --training.seq_len 2048 \
  --training.gradient_accumulation_steps 2 \
  --training.steps ${STEPS} \
  --training.max_norm 1.0 \
  --training.skip_nan_inf \
  --training.dataset ${DATASET} \
  --training.dataset_name ${DATASET_NAME} \
  --training.dataset_split train \
  --training.num_workers 12 \
  --training.prefetch_factor 2 \
  --training.seed 42 \
  --training.compile \
  --checkpoint.interval 2048 \
  --checkpoint.load_step -1 \
  --checkpoint.keep_latest_k 2 \
  --metrics.log_freq 1
```
